stdlib/stack_trace.mli

```diff
@@ -0,0 +1,141 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Stack traces.
+
+    Facilities for getting the current stack trace, printing and inspecting
+    them.
+
+    @since 4.08 *)
+
+type t
+(** The type of stack traces.
+
+    Stack traces cannot be marshalled. If you need marshalling, you should use
+    the array returned by the {!slots} function. *)
+
+val current: int -> t
+(** [current n] returns a description of the stack trace on the current program
+    point (for the current thread), with at most [n] entries. *)
+
+val to_string: t -> string
+(** Return a string listing the program locations contained in the stack
+    trace. *)
+
+val length: t -> int
+(** [length st] returns the number of (non-inlined) slots in the stack trace
+    [st]. *)
+
+module Slot : sig
+  (** {1 Manipulation of stack trace information}
+
+      These functions are used to traverse the slots of a stack trace and
+      extract information from them in a programmer-friendly format. *)
+
+  type stack_trace
+
+  type t
+  (** The abstract type [t] represents a single slot of a stack trace. *)
+
+  val is_raise: t -> bool
+  (** [is_raise slot] is [true] when [slot] refers to a raising point in the
+      code, and [false] when it comes from a simple function call. *)
+
+  val is_inline: t -> bool
+  (** [is_inline slot] is [true] when [slot] refers to a call that got inlined
+      by the compiler, and [false] when it comes from any other context. *)
+
+  type location =
+    { filename: string;
+      line_number: int;
+      start_char: int;
+      end_char: int }
```

Comment:
I wouldn't expose the record fields but rather move it to its own module with a creating function and accessors.

Reply:
Indeed.  Also, I would recommend putting the closing curly brace on its own line (as you have in other places), since it makes it quicker to add another record field.

Reply:
I think it would definitively be useful be able to create values or redefine fields e.g. to remap filenames in a given context.  I would also rename `{start,end}_char` to `{start,end}_byte` since it's likely a better characterization of what it is but that may not be consistent with the rest of the compiler apis.